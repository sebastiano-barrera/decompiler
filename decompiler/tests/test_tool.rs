//! Tests based on "test_tool"
//!
//! These are pretty weak, but they work on real code. They only check that the
//! decompiler doesn't panic, and they write the decompiler's output as text to
//! a version-controlled file, just so that it's somewhere easy to diff and hard
//! to forget updating.

use include_dir::{include_dir, Dir};
use insta::assert_snapshot;

use std::{
    path::Path,
    sync::{Mutex, MutexGuard, OnceLock},
};

static DATA_DIR: Dir<'_> = include_dir!("$CARGO_MANIFEST_DIR/test-data/");

macro_rules! case {
    ($case:ident, $exe:expr, $func_name:ident) => {
        #[test_log::test]
        #[allow(non_snake_case)]
        fn $case() {
            let function_name = stringify!($func_name);
            let out = $exe.process_function(function_name);
            crate::assert_snapshot!(out);
        }
    };
}

macro_rules! tests_in_binary {
    ($group:ident, $exe_path:expr; $($funcs:ident),* $(,)?) => {
        mod $group {
            use crate::Exe;
            static EXE: Exe = Exe::new($exe_path);

            $(case!($funcs, EXE, $funcs);)*
        }
    };
}

struct Exe {
    once_lock: OnceLock<Mutex<decompiler::Executable<'static>>>,
    path: &'static str,
}
impl Exe {
    const fn new(path: &'static str) -> Self {
        Exe {
            once_lock: OnceLock::new(),
            path,
        }
    }

    fn get_or_init(&self) -> MutexGuard<'_, decompiler::Executable<'static>> {
        let mutex = self.once_lock.get_or_init(|| {
            let rel_path = Path::new(self.path);
            let raw = DATA_DIR.get_file(rel_path).unwrap().contents();
            let exe = decompiler::Executable::parse(raw).unwrap();
            Mutex::new(exe)
        });
        mutex.lock().unwrap()
    }

    fn process_function(&self, function_name: &str) -> String {
        use std::fmt::Write;

        let mut exe = self.get_or_init();

        let df = exe
            .decompile_function(function_name)
            .expect("decompiling function");

        let mut log_buf = String::new();

        if let Some(mil) = df.mil() {
            writeln!(log_buf, " --- mil").unwrap();
            writeln!(log_buf, "{:?}\n", mil).unwrap();
        }

        if let Some(ssa) = df.ssa_pre_xform() {
            writeln!(log_buf, " --- ssa pre-xform").unwrap();
            writeln!(log_buf, "{:?}\n", ssa).unwrap();
        }

        if let Some(ssa) = df.ssa() {
            writeln!(log_buf, " --- cfg").unwrap();
            let cfg = ssa.cfg();
            writeln!(log_buf, "  entry: {:?}", cfg.direct().entry_bid()).unwrap();
            for bid in cfg.block_ids() {
                let regs: Vec<_> = ssa.block_regs(bid).collect();
                writeln!(
                    log_buf,
                    "  {:?} -> {:?} {:?}",
                    bid,
                    cfg.block_cont(bid),
                    regs
                )
                .unwrap();
            }
            write!(log_buf, "  domtree:\n    ").unwrap();

            let mut pp_buf = decompiler::pp::FmtAsIoUTF8(&mut log_buf);
            let pp = &mut decompiler::pp::PrettyPrinter::start(&mut pp_buf);
            cfg.dom_tree().dump(pp).unwrap();
            writeln!(log_buf).unwrap();

            writeln!(log_buf, " --- ssa").unwrap();
            writeln!(log_buf, "{:?}\n", ssa).unwrap();

            writeln!(log_buf, " --- ast").unwrap();
            let mut ast = decompiler::Ast::new(&ssa, exe.types());
            let mut pp_buf = decompiler::pp::FmtAsIoUTF8(&mut log_buf);
            let pp = &mut decompiler::pp::PrettyPrinter::start(&mut pp_buf);
            ast.pretty_print(pp).unwrap();
        }

        log_buf
    }
}

tests_in_binary!(
    exe_composite_type, "ty/test_composite_type.so";
    list_len
);
tests_in_binary!(exe_capstone, "integration/sample_capstone"; main);
tests_in_binary!(exe_matrix, "integration/sample_matrix"; sum_matrix);
tests_in_binary!(
    redis_server, "integration/redis-server";
    ctl_lookup,
    listNext,
    listTypePush,
    geoArrayCreate,
    // regression: infinite loop
    processAnnotations,
    redisReconnect,
    redisConnectWithOptions,
);
