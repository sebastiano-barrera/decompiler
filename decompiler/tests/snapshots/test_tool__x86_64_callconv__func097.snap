---
source: decompiler/tests/test_tool.rs
expression: out
---
func097 0x4012ef+7 (file 0x12ef)
          4012ef: f3 0f 10 44 24 10       movss xmm0,[rsp+10h]
          4012f5: c3                      ret

function type: "func097" func (arg0 *void <TypeID(1v1)>,
                               arg1 *void <TypeID(1v1)>,
                               arg2 *void <TypeID(1v1)>,
                               arg3 small <TypeID(785v1)>,
                               arg4 small <TypeID(785v1)>) float <TypeID(783v1)>
0 warnings.


mil program = 
program  89 instrs
    0 r12: TypeID(2v1) <- Ancestral(AncestralName("stack_bottom"))
    1 r2: TypeID(2v1) <- Ancestral(AncestralName("CF"))
    2 r3: TypeID(2v1) <- Ancestral(AncestralName("PF"))
    3 r4: TypeID(2v1) <- Ancestral(AncestralName("AF"))
    4 r5: TypeID(2v1) <- Ancestral(AncestralName("ZF"))
    5 r6: TypeID(2v1) <- Ancestral(AncestralName("SF"))
    6 r7: TypeID(2v1) <- Ancestral(AncestralName("TF"))
    7 r8: TypeID(2v1) <- Ancestral(AncestralName("IF"))
    8 r9: TypeID(2v1) <- Ancestral(AncestralName("DF"))
    9 r10: TypeID(2v1) <- Ancestral(AncestralName("OF"))
   10 r11: TypeID(2v1) <- Ancestral(AncestralName("RBP"))
   11 r12: TypeID(2v1) <- Ancestral(AncestralName("RSP"))
   12 r13: TypeID(2v1) <- Ancestral(AncestralName("RIP"))
   13 r14: TypeID(2v1) <- Ancestral(AncestralName("RDI"))
   14 r15: TypeID(2v1) <- Ancestral(AncestralName("RSI"))
   15 r16: TypeID(2v1) <- Ancestral(AncestralName("RAX"))
   16 r17: TypeID(2v1) <- Ancestral(AncestralName("RBX"))
   17 r18: TypeID(2v1) <- Ancestral(AncestralName("RCX"))
   18 r19: TypeID(2v1) <- Ancestral(AncestralName("RDX"))
   19 r20: TypeID(2v1) <- Ancestral(AncestralName("R8"))
   20 r21: TypeID(2v1) <- Ancestral(AncestralName("R9"))
   21 r22: TypeID(2v1) <- Ancestral(AncestralName("R10"))
   22 r23: TypeID(2v1) <- Ancestral(AncestralName("R11"))
   23 r24: TypeID(2v1) <- Ancestral(AncestralName("R12"))
   24 r25: TypeID(2v1) <- Ancestral(AncestralName("R13"))
   25 r26: TypeID(2v1) <- Ancestral(AncestralName("R14"))
   26 r27: TypeID(2v1) <- Ancestral(AncestralName("R15"))
   27 r28: TypeID(2v1) <- Ancestral(AncestralName("ZMM0"))
   28 r29: TypeID(2v1) <- Ancestral(AncestralName("ZMM1"))
   29 r30: TypeID(2v1) <- Ancestral(AncestralName("ZMM2"))
   30 r31: TypeID(2v1) <- Ancestral(AncestralName("ZMM3"))
   31 r32: TypeID(2v1) <- Ancestral(AncestralName("ZMM4"))
   32 r33: TypeID(2v1) <- Ancestral(AncestralName("ZMM5"))
   33 r34: TypeID(2v1) <- Ancestral(AncestralName("ZMM6"))
   34 r35: TypeID(2v1) <- Ancestral(AncestralName("ZMM7"))
   35 r36: TypeID(2v1) <- Ancestral(AncestralName("ZMM8"))
   36 r37: TypeID(2v1) <- Ancestral(AncestralName("ZMM9"))
   37 r38: TypeID(2v1) <- Ancestral(AncestralName("ZMM10"))
   38 r39: TypeID(2v1) <- Ancestral(AncestralName("ZMM11"))
   39 r40: TypeID(2v1) <- Ancestral(AncestralName("ZMM12"))
   40 r41: TypeID(2v1) <- Ancestral(AncestralName("ZMM13"))
   41 r42: TypeID(2v1) <- Ancestral(AncestralName("ZMM14"))
   42 r43: TypeID(2v1) <- Ancestral(AncestralName("ZMM15"))
   43 r45: TypeID(2v1) <- Undefined
   44 r46: TypeID(2v1) <- Undefined
   45 r47: TypeID(2v1) <- Undefined
   46 r48: TypeID(2v1) <- Undefined
   47 r49: TypeID(2v1) <- Undefined
   48 r50: TypeID(2v1) <- Undefined
   49 r51: TypeID(2v1) <- Undefined
   50 r52: TypeID(2v1) <- Undefined
   51 r53: TypeID(2v1) <- Undefined
   52 r54: TypeID(2v1) <- Undefined
   53 r55: TypeID(2v1) <- Undefined
   54 r56: TypeID(2v1) <- Undefined
   55 r57: TypeID(2v1) <- Undefined
   56 r58: TypeID(2v1) <- Undefined
   57 r59: TypeID(2v1) <- Undefined
   58 r60: TypeID(2v1) <- Undefined
   59 r61: TypeID(2v1) <- Undefined
   60 r62: TypeID(2v1) <- Undefined
   61 r63: TypeID(2v1) <- Undefined
   62 r64: TypeID(2v1) <- Undefined
   63 r65: TypeID(2v1) <- Undefined
   64 r45: TypeID(2v1) <- Ancestral(AncestralName("arg0"))
   65 r15: TypeID(2v1) <- Get(r45)
   66 r46: TypeID(2v1) <- Ancestral(AncestralName("arg1"))
   67 r19: TypeID(2v1) <- Get(r46)
   68 r47: TypeID(2v1) <- Ancestral(AncestralName("arg2"))
   69 r18: TypeID(2v1) <- Get(r47)
   70 r48: TypeID(2v1) <- Ancestral(AncestralName("arg3"))
   71 r49: TypeID(2v1) <- Part { src: r48, offset: 0, size: 8 }
   72 r20: TypeID(2v1) <- Get(r49)
   73 r50: TypeID(2v1) <- Part { src: r48, offset: 8, size: 8 }
   74 r21: TypeID(2v1) <- Get(r50)
   75 r51: TypeID(2v1) <- Ancestral(AncestralName("arg4"))
   76 r53: TypeID(2v1) <- Part { src: r51, offset: 0, size: 8 }
   77 r52: TypeID(2v1) <- ArithK(Add, r12, 8)
   78 r52: TypeID(2v1) <- StoreMem { addr: r52, value: r53 }
   79 r53: TypeID(2v1) <- Part { src: r51, offset: 8, size: 8 }
   80 r52: TypeID(2v1) <- ArithK(Add, r12, 16)
   81 r52: TypeID(2v1) <- StoreMem { addr: r52, value: r53 }
0x4012ef:
   82 r45: TypeID(2v1) <- NotYetImplemented("unsupported: movss xmm0,[rsp+10h]")
0x4012f5:
   83 r45: TypeID(2v1) <- Void
   84 r46: TypeID(2v1) <- Part { src: r28, offset: 0, size: 8 }
   85 r45: TypeID(2v1) <- Concat { lo: r45, hi: r46 }
   86 r45: TypeID(2v1) <- Part { src: r45, offset: 0, size: 4 }
   87 r47: TypeID(2v1) <- SetReturnValue(r45)
   88 r47: TypeID(2v1) <- Control(Ret)


ssa pre-xform:
ssa program  89 instrs
.B0:    ;;  → Always(Return).
  (  2)  r11: ? <- Ancestral(AncestralName("RSP"))
         r27: ? <- Ancestral(AncestralName("ZMM0"))
  (  2)  r75: ? <- Ancestral(AncestralName("arg4"))
         r76: ? <- Part { src: r75, offset: 0, size: 8 }
         r77: ? <- ArithK(Add, r11, 8)
         r78: ? <- StoreMem { addr: r77, value: r76 }
         r79: ? <- Part { src: r75, offset: 8, size: 8 }
         r80: ? <- ArithK(Add, r11, 16)
         r81: ? <- StoreMem { addr: r80, value: r79 }
         r82: ? <- NotYetImplemented("unsupported: movss xmm0,[rsp+10h]")
         r84: ? <- Part { src: r27, offset: 0, size: 8 }
         r85: ? <- Concat { lo: r83, hi: r84 }
         r86: ? <- Part { src: r85, offset: 0, size: 4 }
         r87: ? <- SetReturnValue(r86)


cfg:
  entry: BlockID(0)
  BlockID(0) -> Always(Return) [r11, r27, r75, r76, r77, r78, r79, r80, r81, r82, r83, r84, r85, r86, r87]
  domtree:
    BlockID(0)

ssa post-xform:
ssa program  89 instrs
.B0:    ;;  → Always(Return).
  (  2)  r11: ? <- Ancestral(AncestralName("RSP"))
         r27: ? <- Ancestral(AncestralName("ZMM0"))
  (  2)  r75: ? <- Ancestral(AncestralName("arg4"))
         r76: ? <- Part { src: r75, offset: 0, size: 8 }
         r77: ? <- ArithK(Add, r11, 8)
         r78: ? <- StoreMem { addr: r77, value: r76 }
         r79: ? <- Part { src: r75, offset: 8, size: 8 }
         r80: ? <- ArithK(Add, r11, 16)
         r81: ? <- StoreMem { addr: r80, value: r79 }
         r82: ? <- NotYetImplemented("unsupported: movss xmm0,[rsp+10h]")
         r86: ? <- Part { src: r27, offset: 0, size: 4 }
         r87: ? <- SetReturnValue(r86)



T0: {
  [pre:RSP + 8]:* := pre:arg4[8 .. 0];
  [pre:RSP + 16]:* := pre:arg4[16 .. 8];
  TODO /* unsupported: movss xmm0,[rsp+10h] */;
  ;
  return pre:ZMM0[4 .. 0]
}
